src/app/beer_garden/api/http/handlers/v1/token.py

```diff
@@ -0,0 +1,122 @@
+from beer_garden.errors import ExpiredTokenException, InvalidTokenException
+
+
+class TokenAPI(BaseHandler):
+    def post(self):
+        """
+        ---
+        summary: User login endpoint
+        description: This endpoint is used to do initial authentication via username
+                     and password, which grants an access token to be used on
+                     for authentication against all other protected endpoints, as well
+                     as a refresh token for renewing the access token.
+        parameters:
+          - name: credentials
+            in: body
+            required: false
+            description: The login credentials of the User
+            type: string
+            schema:
+              $ref: '#/definitions/TokenInput'
+        responses:
+          200:
+            description: On successful authentication, a token to be used on subsequent
+                         requests as well as a refresh token for renewing the access
+                         token.
+            schema:
+              $ref: '#/definitions/TokenResponse'
+          401:
+            description: Authentication failed.
+        tags:
+          - Token
+        """
+        user = user_login(self.request)
+
+        if user:
+            response = issue_token_pair(user)
+        else:
+            raise AuthenticationFailed
+
+        self.write(response)
+
+
+class TokenRefreshAPI(BaseHandler):
+    def post(self):
+        """
+        ---
+        summary: Token refresh endpoint
+        description: This endpoint is used to do retrieve a new access token and refresh
+                     token pair using an existing refresh token.
+        parameters:
+          - name: refresh
+            in: body
+            required: true
+            description: A valid refresh token, previously retrieved via either the
+                         /token /token/refresh endpoints.
+            type: string
+            schema:
+              $ref: '#/definitions/TokenRefreshInput'
+        responses:
+          200:
+            description: An access and refresh token pair. The issued pair will replace
+                         the pair with the supplied refresh token's identifier in the
+                         user's list of valid tokens. The expiration time of the new
+                         refresh token will be the same as that of the supplied token.
+            schema:
+              $ref: '#/definitions/TokenResponse'
+          422:
+            description: The provided refresh token is invalid, possibly because it has
+              expired or been revoked.
+        tags:
+          - Token
+        """
+        refresh_token = self.schema_validated_body(TokenRefreshInputSchema)["refresh"]
+
+        try:
+            response = refresh_token_pair(refresh_token=refresh_token)
+        except (ExpiredTokenException, InvalidTokenException):
+            raise UnprocessableEntity(reason="Invalid or expired token")
```

Comment:
Should this really be a 422? [This reference](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422) suggests that error code is for syntactically correct, semantically incorrect errors, particularly (or maybe solely ??) in the context of [WebDAV]( https://datatracker.ietf.org/doc/html/rfc4918#section-11.2).

Whereas [this](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/401) claims that a 401 is for cases where a request lacks valid authentication credentials. To quote [the parent page](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses),

> Although the HTTP standard specifies "unauthorized", semantically this response means "unauthenticated". That is, the client must authenticate itself to get the requested response.

Given that, it feels like the more correct status code is 401.

Reply:
I'll admit that my original implementation had this as a 401, but I eventually realized that was not correct.  My first clue was the UI's behavior when it was a 401, which was to enter an infinite loop of trying to refresh the token.  It was something like:

* Call /api/v1/systems - 401
* I need to authenticate...I have a refresh token, so let's refresh.
* Call /api/v1/token/refresh - 401
* I need to authenticate...I have a refresh token, so let's refresh.
* Repeat forever

Certainly one solution to that could have been adding specific handling for when you get a 401 from the refresh endpoint, but I concluded that that's not correct.  The warning on the Mozilla page for 422 you link to does a good job at explaining why:

> **Warning:** The client should not repeat this request without modification.

A 401 very much indicates that the request you supplied can be retried after authenticating.  You may get a *different* error at that point, but the system is indicating that you can and should retry after authenticating.

That's not the case here and isn't what we want to convey.  The token that the user supplied is not good and authenticating and trying again isn't going to change that fact.  The user in this case **should not retry** the current request as is.  Thus a 422 seems like the appropriate to me.

Reply:
I realize after I wrote all of that that probably one of your objections relates to the fact that 422 is intended as a WebDAV specific response code.  I went back and forth on 400 and 422 for this.  I liked the idea of having two distinct response codes because to me having incorrect input (e.g. a required field is missing) is a different thing than properly formatted input that we reject based on the contents (e.g. you provided a token the way I was expecting, I'm just rejecting said token).  Looking at other frameworks though I do see that it is common to lump those two classes of error under the 400 response code, so I'll go ahead and switch to that instead.